### ISSUE TITLE
Propagating configuration options and scopes in v2

### ISSUE THREAD

**Issue by FSMaxB** [josdejong/mathjs#342]
Right now, configuration options and scopes are only available at parse time. Additionally, scopes and configs are available at compile time due to passing a mathjs instance and scopes (not configuration) are available at evaluation time only if a scope is passed manually.

This makes it impossible to implement configuration options that change the way evaluation works and it makes it impossible to implement configuration options that change the way `toTex` and `toString` work globally without having to pass the options every time one of those functions is called. `toTex` also needs access to the global function scope if we intend to move the LaTeX output of functions to the functions themselves instead of keeping them in a central place.

The easiest solution to this would be to store a reference to the mathjs instance in every node and additionally store a local scope and configuration in every node which has higher precedence than the scope/configuration. This would mean that you can change options globally via the math instance but also override parts of them for a particular node tree.

Those scopes and configuration can then additionally be overridden once more by passing scopes/configuration as argument to the function that is called (`toTex`, `toString`, `eval` ... ).

But @josdejong, in #297 you said: "I'm trying to get rid of passing the global math instance around everywhere, and instead only pass config, and something like dataTypes (containing BigNumber, Matrix, Unit...)".

I don't think it is inherently bad to pass around a reference to a mathjs instance as long as it is forbidden to use it directly but only via wrapper functions that provides access to configuration and scope, we could also provide some kind of encapsulation via a closure to make direct access to the reference impossible ( it would still be accessible from inside the constructor of the node though ). Those wrappers can also handle the case of the referenced global math instance being deleted and can provide defaults in this case.

So to recap: This model has 3 layers of configuration/scope:
* A **global** layer ( the layer of the mathjs instance ). Changing this affects every node tree/parser that got created from this mathjs instance, because those store a reference to this instance.
* A **per node** layer. This layer can be used to override parts or all of the global layer. This could be accessed via a `config` method which makes the configuration propagate down the node tree.
* A **per function call** layer. This layer overrides both the global and the per node layer.

What do you think think of this general concept ( the details can be flashed out later ).

**Comment by josdejong** [comment]
hm, three layers sounds a bit complicated. Maybe we should first get the use cases clear: what would users like to change and configure exactly. What comes in mind here (related to expression parsing) is:

- An option `parenthesis` to configure the way you want to have parenthesis output in `toString` and `toTex`, where possible values could be `'original'`, `'auto'` (only where needed), or `'always'`.

  I'm not sure whether you would like to be able to configure this globally, or per function call, or both. Multiple places to configure the same thing is not always a good idea. Maybe just keep it simple and only pass this option via the functions `toTex(options)` and `toString(options)` (similar to passing output options to `math.format(x, options)`). The current `toTex(callbacks)` may become redundant, read on.

- You would like to be able to define custom `toTex` handlers for new functions. Ideally these are attached to the functions itself, similar to attaching a [transform](http://mathjs.org/docs/expressions/customization.html#function-transforms) to a function. 

- You would like to be able to override the `toTex` methods of individual nodes (maybe you would like the OperatorNode or RangeNode to work differently).

In v1 the nodes don't have access to the global mathjs namespace and/or config, but for v2 we can easily change this, we can put the nodes in a factory, the same way functions are now created in v2:

```js
function factory (type, config, load, typed) {
  function OperatorNode (...) {
    // ...
  }
  // ...
  return OperatorNode;
}

exports.name = 'OperatorNode';
exports.path = 'expression.node';
exports.factory = factory;
```

Some functions (like [eval](https://github.com/josdejong/mathjs/blob/v2/lib/function/expression/eval.js)) needs access to the math namespace containing all functions, in that case you can explicitly request this for your factory function (I agree with you that this is needed in this case):
```js
function factory (type, config, load, typed, math) {
  function OperatorNode (...) {
    // ...
  }
  // ...
  return OperatorNode;
}

exports.name = 'OperatorNode';
exports.path = 'expression.node';
exports.math = true; // request access to the math namespace as 5th argument of the factory function
exports.factory = factory;
```

Such an OperatorNode can be loaded into the math.js namespace using import (same as functions):
```js
math.import(math.require('./expression/node/OperatorNode.js'));
```

I think this new structure would solve a lot of difficulties we had in v1:
- No need to pass `math` to the `Node.compile()` method anymore.
- The nodes can have access to the math namespace when needed, for example FunctionNode and OperatorNode can check whether there is a `toTex` handler attached to a specific function/operator (a function with signature `toTex(node)`?), instead of passing this via config.
- Each math.js instance has a clone of all Nodes, so you can safely override `toTex` inside a specific instance. 
- I think we could retire the `Node.toTex(callbacks: function | Object)` structure then? And instead pass options like `Node.toTex({parenthesis: 'auto'})`.

**Comment by FSMaxB** [comment]
I don't quite grasp this concept in it's entirety but I think it's safe to start by making the nodes use factory functions and flesh out the other details during or after that. Math will be passed only to those nodes who need it.

**Comment by josdejong** [comment]
Yes, let's first refactor the nodes and tackle these things one by one. I expect to have time to refactor the nodes tonight or tomorrow, but if you're faster you could do it too.

**Comment by FSMaxB** [comment]
I don't think I'll have the time today. If you do that: `OperatorNode`, `FunctionNode` and `SymbolNode` need access to the math instance for now.

**Comment by josdejong** [comment]
Ok I've refactored all nodes in to factory functions. `OperatorNode`, `FunctionNode` and `SymbolNode` have access to the math namespace via their factory function.

**Comment by FSMaxB** [comment]
Great, thanks. I've started doing some `toTex` improvements on top of this.

**Comment by FSMaxB** [comment]
I think the current solution actually works pretty well, so I'll close this.
